libs/agno/agno/tools/whatsapp.py

```diff
@@ -0,0 +1,406 @@
+
+        data = {
+            "messaging_product": "whatsapp",
+            "recipient_type": recipient_type,
+            "to": recipient,
+            "type": "text",
+            "text": {"preview_url": preview_url, "body": text},
+        }
+
+        try:
+            response = await self._send_message_async(data)
+            message_id = response.get("messages", [{}])[0].get("id", "unknown")
+            logger.debug(f"Full API response: {json.dumps(response, indent=2)}")
+            return f"Message sent successfully! Message ID: {message_id}"
+        except httpx.HTTPStatusError as e:
+            logger.error(f"Failed to send WhatsApp message: {e}")
+            logger.error(f"Error response: {e.response.text if hasattr(e, 'response') else 'No response text'}")
+            raise
+        except Exception as e:
+            logger.error(f"Unexpected error sending WhatsApp message: {str(e)}")
+            raise
+
+    async def send_template_message_async(
+        self,
+        recipient: Optional[str] = None,
+        template_name: str = "",
+        language_code: str = "en_US",
+        components: Optional[List[Dict[str, Any]]] = None,
+    ) -> str:
+        """Send a template message to a WhatsApp user (asynchronous version).
+
+        Args:
+            recipient: Recipient's WhatsApp ID or phone number (e.g., "+1234567890"). If not provided, uses default_recipient
+            template_name: Name of the template to use
+            language_code: Language code for the template (e.g., "en_US")
+            components: Optional list of template components (header, body, buttons)
+
+        Returns:
+            Success message with message ID
+        """
+        # Use default recipient if none provided
+        if recipient is None:
+            if not self.default_recipient:
+                raise ValueError("No recipient provided and no default recipient set")
+            recipient = self.default_recipient
+
+        logger.debug(f"Sending WhatsApp template message to {recipient}: {template_name}")
+
+        data = {
+            "messaging_product": "whatsapp",
+            "to": recipient,
+            "type": "template",
+            "template": {"name": template_name, "language": {"code": language_code}},
+        }
+
+        if components:
+            data["template"]["components"] = components  # type: ignore[index]
+
+        try:
+            response = await self._send_message_async(data)
+            message_id = response.get("messages", [{}])[0].get("id", "unknown")
+            return f"Template message sent successfully! Message ID: {message_id}"
+        except httpx.HTTPStatusError as e:
+            logger.error(f"Failed to send WhatsApp template message: {e}")
+            raise
+    async def send_image_message_async(self, image: any = None, text: Optional[str] = None, recipient: Optional[str] = None) -> str:
```

Comment:
This isn't really appropriate for an agent to get. Perhaps we should rather make utilities for these and remove them from the Toolkit. I would say the same for template messages really.